### PR TITLE
DDP-5224 cache Elasticsearch clients

### PR DESCRIPTION
**Pull request by yufengwng**
## Context

I saw ES export in `test` environment timeout a few times. I'm not sure root cause, but I realized that the ES clients holds quite a few underlying resources and that we're creating a new client on every batch of participants. Comparing memory usage and "time to complete exports", I didn't notice any major differences at first glance, so I don't have any hard evidence that this helps improve things. However, this intuitively feels like a beneficial thing to do.

## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [x] I have considered error handling and alerts, and added `L-*` labels as needed.
- [x] I have considered security and privacy, and added `I-*` labels as needed
- [x] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.

## FUD Score

_Overall, how are you feeling about these changes?_

- [x] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [x] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [ ] I have written automated positive tests
- [ ] I have written automated negative tests
- [x] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

**Review by ssettipalli — DISMISSED** [review]
looks good

File: pepper-apis/src/main/java/org/broadinstitute/ddp/export/DataExporter.java [hunk 1]
```diff
@@ -140,6 +142,7 @@
     private Gson gson;
     private Set<String> componentNames;
     private PdfService pdfService;
+    private RestHighLevelClient esClient;
```
Comment (zyme) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/export/DataExporter.java]: Maybe make this `final` since `close()` assumes non-null?

**Review by ssettipalli — APPROVED** [review]

File: pepper-apis/src/main/java/org/broadinstitute/ddp/export/DataExporter.java [hunk 2]
```diff
@@ -124,7 +126,7 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-public class DataExporter {
+public class DataExporter implements Closeable {
```
Comment (MocanaAtBroad) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/export/DataExporter.java:129]: Not seeing DataExporter being closed . Intentional?

Comment (MocanaAtBroad) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/export/DataExporter.java:129]: And if you do intend to close() DataExporter, then consider instantiating a DataExporter each time you `run` a Job as opposed to hanging on to the instance time after time and allow GC to reclaim whatever resources are linked back to this.
(Looking at heap dump, I think we are indirectly creating lots of threads that don't  go away. The underlying http client here might be part of this problem).

**Review by MocanaAtBroad — COMMENTED** [review]
Sorry for belated comments!

File: pepper-apis/src/main/java/org/broadinstitute/ddp/export/DataExporter.java (hunk 2 above)
Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/export/DataExporter.java:129]: > Not seeing DataExporter being closed . Intentional?

The DataExporter is mostly used within background jobs that exports data, so those literally hang around for the lifetime of Housekeeping. I'm implicitly letting this get closed by the JVM when Housekeeping gets destroyed.

> consider instantiating a DataExporter each time you run a Job
> ...
> I think we are indirectly creating lots of threads that don't go away

Actually, I want to do the opposite. I don't want to create new DataExporter on every `run` of a job. It's wasteful and I suspect that is causing the `indirectly creating lots of threads that don't go away` that you might have observed. Instead, we're creating the DataExporter just once (or just a few times..) and reusing that.

Comment (MocanaAtBroad) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/export/DataExporter.java:129]: OK. My concern more specifically: the elasticsearch client.
Underneath implementation it creates asynchronous apache http clients.
The worry here is that each elastic search client ends up indirectly creating a separate thread pool. We are creating more threads as we go. Heap analysis is showing me that.
So two ways to protect from building up threads unnecessarily:
1) Go with a true singleton elastic search client (Apparently that is their recommendation) 
https://www.elastic.co/guide/en/elasticsearch/client/java-rest/current/_changing_the_client_8217_s_initialization_code.html
2) Create it use and dispose of it to reduce chance of having multiple ones running at same time.

Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/export/DataExporter.java:129]: Good find! That can be our next step in improvement! I tried digging deeper into the implementation but wasn't too familiar with the async/reactor/NIO stuff. I saw that there was a limit of "max connections" and "max connections per route", so I though there was a upper limit on that. I might be wrong there. I think I'll vote for option 1 of having a singleton.

Comment (MocanaAtBroad) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/export/DataExporter.java:129]: Yeah. That sounds like the better choice.

Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/export/DataExporter.java:129]: Great! I'll create a separate PR for that.

**Pull request merged**